allocator: add an option to reserve additional memory for the OS

wmitros opened the pull request:
This patch adds an option to specify the memory reserved for the OS not as a flat value, but as a modifier for the recommended amounts.

The intended use case of this option is when the application is using both seastar and system allocators and it relies on the seastar defaults for the reservation of memory reserved for the OS but wants to make sure that there is enough memory not managed by seastar allocator for both the application and the rest of the system.

If the user specifies both the reserve-memory and this option, the memory reserved for the system will be a sum of these values. As a result, there will be enough memory for system and non-seastar application allocations, and the seastar allocator will take control of the remaining memory.

Because the optimal value for this option depends on the implementation of the specific application, this option is also added to the app_template config, so that it can be set in the application code.

Signed-off-by: Wojciech Mitros <wojciech.mitros@scylladb.com>

wmitros commented:
This patch will be particularly useful for the mmap allocations added in https://github.com/scylladb/scylladb/pull/11351

avikivity commented on include/seastar/core/app-template.hh:
```diff
@@ -69,6 +69,12 @@ public:
+        /// This can be used when the application allocates some of its memory using the 
+        /// seastar allocator, and some using the system allocator, in particular when it
+        /// uses the mmap system call with MAP_ANONYMOUS which is not overridden in seastar.
+        std::string reserve_additional_memory = "0";
```
Why a string for a number?

xemul commented:
> If the user specifies both the reserve-memory and this option, the memory reserved for the system will be a sum of these values.

Can user specify the reserve-memory as a sum of desired values already? Why adding another option for that?

wmitros commented:
> > If the user specifies both the reserve-memory and this option, the memory reserved for the system will be a sum of these values.
> 
> Can user specify the reserve-memory as a sum of desired values already? Why adding another option for that?

This option is needed for the case where the user does not know how much memory he wants to reserve for the system and prefers to rely on our defaults. 
In the case where both this and the reserve-memory options are used the behavior would be the same as when only the reserve-memory option was used with the sum of those two options, but it would change its meaning - instead of specifying the amount of memory reserved for the OS, it would now mean the sum of the amount of memory for the OS and for the amount of memory allocated using the non-seastar allocator by the app.
The option name may be misleading but, functionally, it describes exactly what it does and I could not come up with a better one

xemul commented:
I see. I would suggest then making those two options mutually exclusive to make them less confusing. So the `--reserve-memory` means -- just take this value and use it and `--extra-reserve-memory` (or maybe even `--reserve-memory-extra`) means take default and add this value on top.

However, thinking about it, a user might want to apply `--reserve-memory-min` option to reserve _at least_ this amount of reserve. This would call for yet another option :confused:

xemul commented:
If the app user knows how many memory the rest of the system needs and doesn't want to mess with the rest, it sounds like `--reserve-memory-min` option, not the `--extra-reserve-memory`

wmitros commented:
Actually, doesn't the `--reserve-memory` option already achieve the same as the `--reserve-memory-min` option? Or are we talking about different options (for now there are `--reserve-memory` for specifying the memory for the system and `--memory` for specifying the amount of memory for the app)
If the app user knows how much memory the rest of the system needs , I expected him to just use `--reserve-memory` option, not the `--extra-reserve-memory`

xemul commented:
> If the app user knows how much memory the rest of the system needs , I expected him to just use --reserve-memory option, not the --extra-reserve-memory

Then I withdraw my "I see" from the above :( because I no longer understand the use-case for the `--extra-reserve-memory` option.

> This option is needed for the case where the user does not know how much memory he wants to reserve for the system and prefers to rely on our defaults.

If he doesn't know how much memory he wants to reserve for the system, what would be his choice of `--extra-reserve-memory` value then?

wmitros commented:
Maybe I tried to make this option too general. 

The issue that this option helps to solve is when an app uses both the seastar allocator (which allows allocating from a pool of preallocated memory which size can be set using `--memory`) and the mmap system calls with MAP_ANONYMOUS - which allocate memory from the pool reserved for the system, which size can be set using `--reserve-memory`. 

Until now, the seastar memory was synonymous with the memory used by the app, if we used  `--reserve-memory=x`, we knew that there will always be `x` memory for the system. Now if we use only `--reserve-memory=x`, there may be only x-[memory from mmaps used by the app]. With the new option, the app could specify how much of memory that it allocates comes from mmaps (we don't really expect the user to take this option into account)

xemul commented:
This:

> the mmap system calls with MAP_ANONYMOUS - which allocate memory from the pool reserved for the system

and

> Now if we use only --reserve-memory=x, there may be only x-[memory from mmaps used by the app]

Sounds like some time before mmap()-ing the memory was prohibited (or simply didn't happen), but now some place(s) in code started calling mmap(), not seastar allocator, to get memory. Is that correct? If yes, what are those mmap() callers?

avikivity commented:
> This:
> 
> > the mmap system calls with MAP_ANONYMOUS - which allocate memory from the pool reserved for the system
> 
> and
> 
> > Now if we use only --reserve-memory=x, there may be only x-[memory from mmaps used by the app]
> 
> Sounds like some time before mmap()-ing the memory was prohibited (or simply didn't happen), but now some place(s) in code started calling mmap(), not seastar allocator, to get memory. Is that correct? If yes, what are those mmap() callers?

The WASM compiler and runtime.

(The runtime reserves 4G of address space to avoid pointer checks)

xemul commented:
> The WASM compiler and runtime.
> (The runtime reserves 4G of address space to avoid pointer checks)

Then it's scylla-specific (at least for now). It's worth making this option programmatically-configurable only and marking as "experimental" or "internal"

wmitros commented:
I agree that it makes most sense making it programmatically-configurable only. I rebased with the applied change.
As a side effect, there is now no reason for the `app_template::config::reserve_additional_memory` to be a string, so now it's an integer.

I also modified the behaviour of the `total_memory` option. Now we subtract the extra memory from the total app memory so the actual memory used by scylla is still limited by this option

wmitros commented:
@avikivity @xemul do you think the patch looks better now after the last change?

xemul commented on src/core/resource.cc:
```diff
@@ -212,14 +212,15 @@ namespace resource {
         available_memory = min_memory;
     }
-    size_t mem = c.total_memory.value_or(available_memory);
+    size_t mem = c.total_memory ? *c.total_memory - c.reserve_additional_memory : available_memory;
```
Why only substract reserve_additional_memory from explicitly configured memory? If seastar is allowed to use all it wants (i.e. -- no -m is given), why does it make any difference in terms of

> Now we subtract the extra memory from the total app memory so the actual memory used by scylla is still limited by this option

?

wmitros replied:
We subtract the extra memory in both cases, whether `-m` was given or not (we want this `mem` variable here to be lower) - the subtraction happens either here in line 223 (if `-m` is given) or as a result of lines 215 and 218 (if `-m` is not given). 
I omitted the low memory configuration case (the 500MB) because in that case we aren't obeying our specified `--reserve-memory` anyway, so we aren't breaking any assumptions by leaving even less memory for the system and if we assume that an app may need 500MB for the seastar allocations, it will probably keep needing it. But if it makes more sense to reserve the memory in this case as well, we can do it.

xemul replied:
> We subtract the extra memory in both cases, whether -m was given or not (we want this mem variable here to be lower) - the subtraction happens either here in line 223 (if -m is given) or as a result of lines 215 and 218 (if -m is not given).

Ah, indeed :exploding_head:

xemul commented:
LGTM from my side

wmitros commented:
@avikivity we got one LGTM already but you can also make sure that there are no more comments from your side
